sql-plugin/src/main/scala/com/nvidia/spark/rapids/limit.scala

```diff
@@ -248,54 +248,72 @@ object GpuTopN {
       inputRows: GpuMetric,
       outputBatches: GpuMetric,
       outputRows: GpuMetric,
-      offset: Int): Iterator[ColumnarBatch] =
-    new Iterator[ColumnarBatch]() {
+      offset: Int): Iterator[SpillableColumnarBatch] =
+    new Iterator[SpillableColumnarBatch]() {
       override def hasNext: Boolean = iter.hasNext
 
       private[this] var pending: Option[SpillableColumnarBatch] = None
 
-      override def next(): ColumnarBatch = {
-        while (iter.hasNext) {
-          val input = iter.next()
-          withResource(new NvtxWithMetrics("TOP N", NvtxColor.ORANGE, opTime)) { _ =>
-            inputBatches += 1
-            inputRows += input.numRows()
-            lazy val totalSize = GpuColumnVector.getTotalDeviceMemoryUsed(input) +
-                pending.map(_.sizeInBytes).getOrElse(0L)
-
-            val runningResult = if (pending.isEmpty) {
-              withResource(input) { input =>
-                apply(limit, sorter, input, sortTime)
-              }
-            } else if (totalSize > Integer.MAX_VALUE) {
-              // The intermediate size is likely big enough we don't want to risk an overflow,
-              // so sort/slice before we concat and sort/slice again.
-              val tmp = withResource(input) { input =>
-                apply(limit, sorter, input, sortTime)
-              }
-              withResource(concatAndClose(pending.get, tmp, concatTime)) { concat =>
-                apply(limit, sorter, concat, sortTime)
-              }
-            } else {
-              // The intermediate size looks like we could never overflow the indexes so
-              // do it the more efficient way and concat first followed by the sort/slice
-              withResource(concatAndClose(pending.get, input, concatTime)) { concat =>
-                apply(limit, sorter, concat, sortTime)
+      override def next(): SpillableColumnarBatch = {
+        if (!hasNext) {
+          throw new NoSuchElementException()
+        }
+        closeOnExcept(pending) { _ =>
+          while (iter.hasNext) {
+            val inputScb = closeOnExcept(iter.next()) { cb =>
+              inputBatches += 1
+              inputRows += cb.numRows()
+              SpillableColumnarBatch(cb, SpillPriorities.ACTIVE_ON_DECK_PRIORITY)
+            }
+            withResource(new NvtxWithMetrics("TOP N", NvtxColor.ORANGE, opTime)) { _ =>
+              withRetry(inputScb, splitSpillableInHalfByRows) { attempt =>
```

Comment:
nit: can we move this outside of the metrics range? It just means that we are more likely to capture the time spent doing computation instead of the time spent with the thread paused waiting to retry.